blog/2020-11-19-error-message-design.md

```diff
@@ -0,0 +1,62 @@
+
+We must acknowledge two things:
+
+- there is immense information asymmetry between framework developer and framework user (developers know the codebase far better than users)
+- error messages are part of your developer experience too
+
+You might be familiar with the old adage "If you fail to plan, you plan to fail." Here, we have a slight twist: **If you fail to plan to fail, you are planning to fail your users when they need you most.** Touting great developer experience with lousy error message design is like selling a beautiful car with seatbelts that don't work when there's a crash.
+
+## Error Message Style Guide
+
+At Temporal, our current style guide reflects our engineering roots in [Uber](https://github.com/uber-go/guide/blob/master/style.md#error-types) and [Golang](https://blog.golang.org/go1.13-errors) with some tactical recommendations:
+
+- Use `errors.New` for simple messages and `fmt.Errorf` for formatted messages. Make sure `errors` is standard go package.
+- Error messages start with lowercase and no punctuation at the end: `unable to open file` instead of `Unable to open file..`
+- Error messages should be actionable, e.g. `no such directory` instead of `loading service config`.
+- Use `unable to...` instead of `failed to...`, `can't...`, or `could not...`.
+- Messages should reflect logic action that we just tried to make (in a call on previous line which returned an error), not a function name: `load service config` instead of `LoadServiceConfig`. This message should add actionable meaningful context to the existing error message.
+
+That said, as we look to evolve our philosophy of developer experience, we are looking toward higher level design principles that help guide successful outcomes.
+
+## Error Message Design Principles 
+
+We are still in the early stages of pinning down our beliefs in what makes for great error messages, but here are some candidate principles running through my mind:
+
+- **Errors should be written for USERS, not maintainers**: Your error should make sense from a user point of view. For example, you should cite user-supplied configs and values familiar to users, not the internal variable and function names familiar to maintainers. 
+- **Errors should suggest WHY they happened**, if not HOW to fix them: As a maintainer, you probably have a high amount of context as to the possible causes of an error - take the time to write down some hints for your users. If there's a high probability you know what's wrong, the error message itself can suggest how to fix it. But even without that certainty, think about what the user is likely to need to log out to debug the error, and offer it in the error message itself. Where relevant, **Errors should display what was expected vs what was received**.
```

Comment:
```suggestion
- **Errors should suggest WHY they happened**, or even better - HOW to fix them: As a maintainer, you probably have a high amount of context as to the possible causes of an error - take the time to write down some hints for users. If there's a high probability you know what's wrong, the error message itself can suggest how to fix it. But even without that certainty, think about what the user is likely to need to log out to debug the error, and offer it in the error message itself. Where relevant, **Errors should display what was expected vs what was received**.
```
" think about what the user is likely to need to log out to debug the error, and offer it in the error message itself."

English is wrong I think